editoast/src/modelsv2/rolling_stock_model/schedules_from_rolling_stock.rs

```diff
@@ -0,0 +1,112 @@
+use diesel::sql_query;
+use diesel::sql_types::{BigInt, Nullable, Text};
+use diesel_async::RunQueryDsl;
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+use crate::error::Result;
+use editoast_models::db_connection_pool::DbConnectionV2;
+
+#[derive(Debug, Serialize, Deserialize, ToSchema)]
+pub struct RollingStockUsage {
+    #[schema(inline)]
+    pub train_schedule: TrainScheduleReference,
+    #[schema(inline)]
+    pub op_study: Option<StudyReference>,
+}
+
+#[derive(Debug, Serialize, Deserialize, ToSchema)]
+pub struct TrainScheduleReference {
+    pub id: i64,
+    pub name: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, ToSchema)]
+pub struct StudyReference {
+    pub project_id: i64,
+    pub project_name: String,
+    pub study_id: i64,
+    pub study_name: String,
+    pub scenario_id: i64,
+    pub scenario_name: String,
+}
+
+#[derive(QueryableByName, Debug)]
+struct SchedulesFromRollingStock {
+    #[diesel(sql_type = BigInt)]
+    id: i64,
+    #[diesel(sql_type = Text)]
+    name: String,
+    #[diesel(sql_type = Nullable<BigInt>)]
+    project_id: Option<i64>,
+    #[diesel(sql_type = Nullable<Text>)]
+    project_name: Option<String>,
+    #[diesel(sql_type = Nullable<BigInt>)]
+    study_id: Option<i64>,
+    #[diesel(sql_type = Nullable<Text>)]
+    study_name: Option<String>,
+    #[diesel(sql_type = Nullable<BigInt>)]
+    scenario_id: Option<i64>,
+    #[diesel(sql_type = Nullable<Text>)]
+    scenario_name: Option<String>,
+}
+
+impl From<SchedulesFromRollingStock> for RollingStockUsage {
+    fn from(value: SchedulesFromRollingStock) -> Self {
+        let SchedulesFromRollingStock {
+            id,
+            name,
+            project_id,
+            project_name,
+            study_id,
+            study_name,
+            scenario_id,
+            scenario_name,
+        } = value;
+        let study_reference = match (
+            project_id,
+            project_name,
+            study_id,
+            study_name,
+            scenario_id,
+            scenario_name,
+        ) {
+            (
+                Some(project_id),
+                Some(project_name),
+                Some(study_id),
+                Some(study_name),
+                Some(scenario_id),
+                Some(scenario_name),
+            ) => Some(StudyReference {
+                project_id,
+                project_name,
+                study_id,
+                study_name,
+                scenario_id,
+                scenario_name,
+            }),
+            _ => None,
+        };
```

Comment:
Admittedly it's not much better. I like it slightly more as it spreads over less lines but that's aesthetics and it forces us to deal with nested tuples. I'm not sure there's a cleaner standard way. We have [itertools](https://docs.rs/itertools/latest/itertools/) already in our deps, maybe something there would be useful? (`Option::iter` returns an iterator with 0 or 1 value.)

In any case, `zip` was a suggestion more than anything else, just keep the version you prefer :)

Reply:
Maybe https://docs.rs/itertools/latest/itertools/macro.izip.html?

Reply:
It feels like `Option` could use some sort of `zip_multiple` macro for such cases that would allow to build from `n` options in one go an option containing a n-tuple equivalent to calling `Option::zip` n times and then flattening the result.

I agree the `Option::zip` approach is a bit more concise, I will go with it. Thanks for mentioning `Option::zip`, I did not know about it !

Reply:
FYI there's one there:

https://github.com/OpenRailAssociation/osrd/blob/b4cfac0e0659d6cbe3735d32d919eeba68500b13/editoast/editoast_derive/src/modelv2/utils.rs#L4-L17

However I'm not in favor of using it in editoast itself as it's both non-standard and not coming from a "popular" crate. (Ie. let's not invent our own stdlib.) It's useful in that module though because 1. the manual alternative is (much) worse, 2. the macro is only involved locally, 3. it's used **extensively**.

Reply:
Maybe something like this?

```rust
let scenario_reference = project_id
    .and_then(|project_id| Some((project_id, project_name?)))
    .and_then(|(project_id, project_name)| Some((project_id, project_name, study_id?)))
    .and_then(|(project_id, project_name, study_id)| Some((project_id, project_name, study_id, study_name?)))
    // yadiyadiyada
    .and_then(|(project_id, project_name, study_id, study_name, scenario_id, scenario_name)| ScenarioReference {
        project_id,
        project_name,
        study_id,
        study_name,
        scenario_id,
        scenario_name,
    })
```

Reply:
That's much better indeed :) I'm not sure about the second part though, `and_then` is defined like:

```rust
    pub fn and_then<U, F>(self, f: F) -> Option<U>
    where
        F: FnOnce(T) -> Option<U>,
    {
        match self {
            Some(x) => f(x),
            None => None,
        }
    }
```

Reply:
Thank you for your comments !

I also find the `project_id.and_then(...)` option to be the cleanest one. I will go with it by default.

Reply:
Note, after discussion with @leovalais, `None.and_then(|()| todo!())` will not execute the closure, but `Some(()).and_then(|_| todo!())`. Maybe we want to write it like this?

Reply:
We should probably just call the closure ourselves...

```rust
let scenario_reference = (|| Some(ScenarioReference {
        project_id: project_id?,
        project_name: project_name?,
        study_id: study_id?,
        study_name: study_name?,
        scenario_id: scenario_id?,
        scenario_name: scenario_name?,
    }))();
```

Reply:
![cease-fire](https://github.com/user-attachments/assets/fdcce2da-9a0e-415d-befd-cfc5d4ffe9c6)